src/widget/wtrackmenu.cpp

```diff
@@ -0,0 +1,1434 @@
+    m_pTrackInfo->loadTrack(pTrack); // NULL is fine.
+    currentTrackInfoIndex = index;
+    m_pTrackInfo->show();
+}
+
+void WTrackMenu::showTrackInfo(TrackPointer pTrack) {
+    if (m_pTrackInfo.isNull()) {
+        m_pTrackInfo.reset(new DlgTrackInfo(m_pConfig, nullptr));
+        connect(m_pTrackInfo.data(), &DlgTrackInfo::showTagFetcher, this, &WTrackMenu::slotShowTrackInTagFetcher);
+        connect(m_pTrackInfo.data(), &DlgTrackInfo::finished, this, &WTrackMenu::slotTrackInfoClosed);
+    }
+    m_pTrackInfo->loadTrack(pTrack); // NULL is fine.
+    m_pTrackInfo->show();
+}
+
+void WTrackMenu::slotNextDlgTagFetcher() {
+    auto trackModel = getTrackModel();
+    if (!trackModel) {
+        return;
+    }
+    QModelIndex nextRow = currentTrackInfoIndex.sibling(
+            currentTrackInfoIndex.row() + 1, currentTrackInfoIndex.column());
+    if (nextRow.isValid()) {
+        showDlgTagFetcher(nextRow);
+        if (!m_pTrackInfo.isNull()) {
+            showTrackInfo(nextRow);
+        }
+    }
+}
+
+void WTrackMenu::slotPrevDlgTagFetcher() {
+    auto trackModel = getTrackModel();
+    if (!trackModel) {
+        return;
+    }
+    QModelIndex prevRow = currentTrackInfoIndex.sibling(
+            currentTrackInfoIndex.row() - 1, currentTrackInfoIndex.column());
+    if (prevRow.isValid()) {
+        showDlgTagFetcher(prevRow);
+        if (!m_pTrackInfo.isNull()) {
+            showTrackInfo(prevRow);
+        }
+    }
+}
+
+void WTrackMenu::showDlgTagFetcher(QModelIndex index) {
+    auto trackModel = getTrackModel();
+    VERIFY_OR_DEBUG_ASSERT(trackModel) {
+        return;
+    }
+
+    TrackPointer pTrack = trackModel->getTrack(index);
+    currentTrackInfoIndex = index;
+    slotShowTrackInTagFetcher(pTrack);
+}
+
+void WTrackMenu::showDlgTagFetcher(TrackPointer pTrack) {
+    slotShowTrackInTagFetcher(pTrack);
+}
+
+void WTrackMenu::slotShowTrackInTagFetcher(TrackPointer pTrack) {
+    if (m_pTagFetcher.isNull()) {
+        bool trackNavigation = getTrackModel() != nullptr;
+        m_pTagFetcher.reset(new DlgTagFetcher(nullptr, trackNavigation));
+        connect(m_pTagFetcher.data(), &DlgTagFetcher::next, this, &WTrackMenu::slotNextDlgTagFetcher);
+        connect(m_pTagFetcher.data(), &DlgTagFetcher::previous, this, &WTrackMenu::slotPrevDlgTagFetcher);
+        connect(m_pTagFetcher.data(), &DlgTagFetcher::finished, this, &WTrackMenu::slotTagFetcherClosed);
+    }
+
+    // NULL is fine
+    m_pTagFetcher->loadTrack(pTrack);
+    m_pTagFetcher->show();
+}
+
+void WTrackMenu::slotShowDlgTagFetcher() {
+    auto trackModel = getTrackModel();
+    if (trackModel) {
+        const auto indices = getTrackIndices();
+        if (indices.size() > 0) {
+            showDlgTagFetcher(indices.at(0));
+        }
+    } else {
+        const auto trackPointers = getTrackPointers();
+        if (trackPointers.size() > 0) {
+            showDlgTagFetcher(trackPointers.at(0));
+        }
+    }
+}
```

Comment:
When moving the next/previous implementation into DlgTagFetcher and DlgTrackInfo, these checks if there is a TrackModel* and different code branches in WTrackMenu can be removed; just pass the TrackPointer to the dialog.

Reply:
These checks will certainly be removed by implementing the self-contained track dialogs.